src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts

```diff
@@ -163,9 +179,50 @@ export class AddRuleComponent implements OnInit, OnDestroy {
     }
 
     add() {
-        this.close();
+        if (this.rule.scope_selectors.repository[0].decoration !== "repoMatches"
+        && this.rule.scope_selectors.repository[0].pattern.indexOf("**") !== -1) {
+            this.inlineAlert.showInlineError(INVALID_RULE);
+            return;
+        }
+        if (this.hasParam()
+          && (this.rule.params[this.template] <= 0 || this.rule.params[this.template] > MAX)) {
+            this.inlineAlert.showInlineError(ILLEGAL_RULE);
+            return;
+        }
+        if (this.isExistingRule()) {
+            this.inlineAlert.showInlineError(EXISTING_RULE);
+            return;
+        }
         this.clickAdd.emit(this.rule);
     }
+    isExistingRule(): boolean {
+        if (this.retention && this.retention.rules && this.retention.rules.length > 0) {
+            for (let i = 0; i < this.retention.rules.length; i++) {
+                if (this.isSameRule(this.retention.rules[i])) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    isSameRule(rule: Rule): boolean {
```

Comment:
Why not just compare the rule object are equal? Is there some property different?

Reply:
according to the design, some key properties are equal, then we view these two rules are the same